File: data/src/main/java/com/somename/data/repository/datasource/mapper/ProductEntityMapper.java
package com.somename.data.repository.datasource.mapper;

import com.somename.data.сontent.ProductEntity;
import com.somename.domain.model.Product;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class ProductEntityMapper extends Mapper<Product, ProductEntity> {

    private final ScreeenshotUrlEntityMapper mScreeenshotUrlEntityMapper;

    @Inject
    public ProductEntityMapper(ScreeenshotUrlEntityMapper screeenshotUrlEntityMapper) {
        mScreeenshotUrlEntityMapper=screeenshotUrlEntityMapper;
    }

    @Override
    public ProductEntity map(Product value) {
        throw new UnsupportedOperationException();
    }

    @Override
    public Product reverseMap(ProductEntity productEntity) {
        Product product = new Product();
        product.setId(productEntity.getId());
        product.setName(productEntity.getName());
        product.setRedirectUrl(productEntity.getRedirectUrl());
        product.setScreenshotUrl(mScreeenshotUrlEntityMapper.reverseMap(productEntity.getScreenshotUrl()));
        product.setTagline(productEntity.getTagline());
        product.setVotesCount(productEntity.getVotesCount());
        return product;
    }
}

File: data/src/main/java/com/somename/data/repository/datasource/mapper/ProductEntityMapper.kt
package com.somename.data.repository.datasource.mapper

import com.somename.data.сontent.ProductEntity
import com.somename.domain.model.Product

import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class ProductEntityMapper @Inject
constructor(private val mScreenshotUrlEntityMapper: ScreenshotUrlEntityMapper) : Mapper<Product, ProductEntity>() {

    override fun map(value: Product): ProductEntity {
        throw UnsupportedOperationException()
    }

    override fun reverseMap(value: ProductEntity): Product {
        val product = Product()
        product.id = value.id
        product.name = value.name
        product.redirectUrl = value.redirectUrl
        product.screenshotUrl = value.screenshotUrl?.let { mScreenshotUrlEntityMapper.reverseMap(it) }
        product.tagline = value.tagline
        product.votesCount = value.votesCount
        return product
    }
}
